File: app/backend/schemas/progress.py
"""Progress tracking schemas"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from app.backend.models.progress import ProgressStatus


class ProgressBase(BaseModel):
    """Base progress schema"""
    status: ProgressStatus
    completion_percentage: float = Field(..., ge=0.0, le=100.0)


class ProgressCreate(ProgressBase):
    """Progress creation schema"""
    module_id: int


class ProgressUpdate(BaseModel):
    """Progress update schema"""
    status: Optional[ProgressStatus] = None
    completion_percentage: Optional[float] = Field(None, ge=0.0, le=100.0)


class ProgressResponse(ProgressBase):
    """Progress response schema"""
    id: int
    user_id: int
    module_id: int
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    last_accessed_at: datetime
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class ProgressListResponse(BaseModel):
    """List of user progress"""
    progress: list[ProgressResponse]
    total: int

File: app/frontend/src/types/progress.ts
/** Progress tracking types */
export type ProgressStatus = "not_started" | "in_progress" | "completed";

export interface Progress {
  id: number;
  user_id: number;
  module_id: number;
  status: ProgressStatus;
  completion_percentage: number;
  started_at: string | null;
  completed_at: string | null;
  last_accessed_at: string;
  created_at: string;
  updated_at: string | null;
}

export interface ProgressListResponse {
  progress: Progress[];
  total: number;
}

export interface ProgressUpdate {
  status?: ProgressStatus;
  completion_percentage?: number;
}

